src/main/java/ru/ralsei/whatcanyousee/logic/LeverGameMap.java

```diff
@@ -0,0 +1,190 @@
+    }
+
+    protected void setPressSelf() {
+        pressSelf = true;
+    }
+
+    /**
+     * Class representing the state of the lever game. All states represents the vertexes of a directed
+     * graph, with levers representing the edges.
+     */
+    public abstract class State {
+        /**
+         * Image to show on this state.
+         */
+        private int imageID;
+
+        /**
+         * True if enters this state meaning winning the game.
+         */
+        private boolean winState = false;
+
+        /**
+         * Message to show if this is a lose state.
+         */
+        private String message;
+
+        String getMessage() {
+            return message;
+        }
+
+        /**
+         * True if enters this state meaning loosing the game.
+         */
+        private boolean loseState = false;
+
+        protected State(int imageID) {
+            this.imageID = imageID;
+        }
+
+        public int getImageID() {
+            return imageID;
+        }
+
+        public void setImageID(int imageID) {
+            this.imageID = imageID;
+        }
+
+        boolean isWinState() {
+            return winState;
+        }
+
+        public void setWinState(boolean winState) {
+            this.winState = winState;
+        }
+
+        boolean isLoseState() {
+            return loseState;
+        }
+
+        public void setLoseState(boolean loseState) {
+            this.loseState = loseState;
+        }
+
+        public void setMessage(String message) {
+            this.message = message;
+        }
+    }
+
+    public LeverGameMap(GameActivity activity) {
+        this.activity = activity;
+    }
+
+    /**
+     * Handles the pressed lever (by the other player).
+     */
+    public abstract void applyLever(String leverName);
+
+    public State getCurrentState() {
+        return states.get(currentStateNumber);
+    }
+
+    public int getCurrentStateNumber() {
+        return currentStateNumber;
+    }
+
+    public int getStateNumber(State currentState) {
+        for (int i = 0; i < states.size(); i++) {
+            if (states.get(i) == currentState) {
+                return i;
+            }
+        }
+
+        return -1;
+    }
+
+    protected void setCurrentState(final int currentStateNumber) {
+        this.currentStateNumber = currentStateNumber;
+
+        activity.runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                ImageView imageView = activity.findViewById(R.id.leverImage);
+                if (imageView != null) {
+                    imageView.setImageResource(LeverGameMap.this.getCurrentState().getImageID());
+                }
+            }
+        });
+
+        if (states.get(currentStateNumber).isWinState()) {
+            activity.runOnUiThread(new Runnable() {
+                @Override
+                public void run() {
+                    activity.getGameplayHandler().onLeverGameWon();
```

Comment:
It's not a good idea to do any blocking operations on UI thread, and this code sends a network message, so potentially blocks. Have you considered using AsyncTask for this kind of operations?

Reply:
RealTimeMultiplayerClient.sendReliableMessage() creates and runs asynchronous task itself
https://developers.google.com/android/reference/com/google/android/gms/games/RealTimeMultiplayerClient
So there is no problem with blocks here.